src/test/mir-opt/early_otherwise_branch.opt1.EarlyOtherwiseBranch.diff

```diff
@@ -0,0 +1,88 @@
+      let mut _4: std::option::Option<usize>; // in scope 0 at $DIR/early_otherwise_branch.rs:6:12: 6:13
+      let mut _5: std::option::Option<usize>; // in scope 0 at $DIR/early_otherwise_branch.rs:6:14: 6:15
+      let mut _6: isize;                   // in scope 0 at $DIR/early_otherwise_branch.rs:7:19: 7:26
+      let mut _7: isize;                   // in scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
+      let _8: usize;                       // in scope 0 at $DIR/early_otherwise_branch.rs:7:15: 7:16
+      let _9: usize;                       // in scope 0 at $DIR/early_otherwise_branch.rs:7:24: 7:25
++     let mut _10: isize;                  // in scope 0 at $DIR/early_otherwise_branch.rs:7:19: 7:26
++     let mut _11: bool;                   // in scope 0 at $DIR/early_otherwise_branch.rs:7:19: 7:26
+      scope 1 {
+          debug a => _8;                   // in scope 1 at $DIR/early_otherwise_branch.rs:7:15: 7:16
+          debug b => _9;                   // in scope 1 at $DIR/early_otherwise_branch.rs:7:24: 7:25
+      }
+  
+      bb0: {
+          StorageLive(_3);                 // scope 0 at $DIR/early_otherwise_branch.rs:6:11: 6:16
+          StorageLive(_4);                 // scope 0 at $DIR/early_otherwise_branch.rs:6:12: 6:13
+          _4 = _1;                         // scope 0 at $DIR/early_otherwise_branch.rs:6:12: 6:13
+          StorageLive(_5);                 // scope 0 at $DIR/early_otherwise_branch.rs:6:14: 6:15
+          _5 = _2;                         // scope 0 at $DIR/early_otherwise_branch.rs:6:14: 6:15
+          (_3.0: std::option::Option<usize>) = move _4; // scope 0 at $DIR/early_otherwise_branch.rs:6:11: 6:16
+          (_3.1: std::option::Option<usize>) = move _5; // scope 0 at $DIR/early_otherwise_branch.rs:6:11: 6:16
+          StorageDead(_5);                 // scope 0 at $DIR/early_otherwise_branch.rs:6:15: 6:16
+          StorageDead(_4);                 // scope 0 at $DIR/early_otherwise_branch.rs:6:15: 6:16
+          _7 = discriminant((_3.0: std::option::Option<usize>)); // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
+-         switchInt(move _7) -> [1_isize: bb2, otherwise: bb1]; // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
++         StorageLive(_10);                // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
++         _10 = discriminant((_3.1: std::option::Option<usize>)); // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
++         StorageLive(_11);                // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
++         _11 = Ne(_10, _7);               // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
++         StorageDead(_10);                // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
++         switchInt(move _11) -> [false: bb6, otherwise: bb1]; // scope 0 at $DIR/early_otherwise_branch.rs:7:10: 7:17
```

Comment:
instead of doing a switch and going to a block that does another switch, you can also compute another comparison:

```rust
_42 = Eq(_7, 1_isize)
_43 = BitAnd(_42, _11)
```

and then switch on `_43` to directly go to `bb3`

Reply:
@oli-obk  Hmm that works when we have only have 2 targets in the switch, but what if we have 3 targets in the switch like `match (x,y,z)`? See the 3_element_tuple test. Correct me if I am wrong though

Reply:
That test makes `bb2` dead code, and doesn't affect the match part on `z` at all, so it all seems to work out to me? You aren't optimizing chains longer than 2 elements with your optimization, and even if the optimization is extended to support that, we'd just end up with more of these `BitAnd`s I think.

Reply:
It's been a while, but I finally looked at it again. I don't think it is that easy to use BitAnd to avoid the extra bb. See https://github.com/rust-lang/rust/pull/75119/files#diff-37f80fe71480f2b22fb8fd1fc103fda4R88 where we need to check both 0 and 1.